game/tests.py

```
@@ -69,6 +71,63 @@ def test_falls_back_to_python_engine_script(self):
                 [sys.executable, candidates[2]],
             )
 
+
+class EngineValidateLegalityTest(SimpleTestCase):
+    """Direct engine validation should reject self-check moves."""
+
+    def _run_python_engine(self, command):
```

Ruff (0.15.13)

[warning] Missing return type annotation for private function `_run_python_engine`

(ANN202)